core/src/main/scala/kafka/log/Log.scala

```diff
@@ -181,29 +190,53 @@ class Log(val dir: File,
         // if its a log file, load the corresponding log segment
         val start = filename.substring(0, filename.length - LogFileSuffix.length).toLong
         val indexFile = Log.indexFilename(dir, start)
-        val segment = new LogSegment(dir = dir,
-                                     startOffset = start,
-                                     indexIntervalBytes = config.indexInterval,
-                                     maxIndexSize = config.maxIndexSize,
-                                     rollJitterMs = config.randomSegmentJitter,
-                                     time = time,
-                                     fileAlreadyExists = true)
-
-        if(indexFile.exists()) {
-          try {
+        val timeIndexFile = Log.timeIndexFilename(dir, start)
+        val loadedSegment = {
+          if (indexFile.exists()) {
+            val segment = new LogSegment(dir = dir,
+              startOffset = start,
+              indexIntervalBytes = config.indexInterval,
+              maxIndexSize = config.maxIndexSize,
+              rollJitterMs = config.randomSegmentJitter,
+              time = time,
+              fileAlreadyExists = true)
+
+            try {
               segment.index.sanityCheck()
-          } catch {
-            case e: java.lang.IllegalArgumentException =>
-              warn("Found a corrupted index file, %s, deleting and rebuilding index...".format(indexFile.getAbsolutePath))
-              indexFile.delete()
-              segment.recover(config.maxMessageSize)
+            } catch {
+              case e: java.lang.IllegalArgumentException =>
+                warn(s"Found a corrupted index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                  s"${indexFile.getAbsolutePath} and rebuilding index...")
+                indexFile.delete()
+                timeIndexFile.delete()
+                segment.recover(config.maxMessageSize)
+            }
+
+            try {
+              segment.timeIndex.sanityCheck()
+            } catch {
+              case e: java.lang.IllegalArgumentException =>
+                warn(s"Found a corrupted time index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                  s"${indexFile.getAbsolutePath} and rebuilding index...")
+                indexFile.delete()
+                timeIndexFile.delete()
+                segment.recover(config.maxMessageSize)
```

Comment:
Do we need to duplicate the code for index and timeindex? Since both indexes are of AbstractIndex, perhaps we can just iterate two AbstractIndex in a loop and do the sanityCheck and recover in the loop (we may need to add a absolutePath() method in AbstractIndex).